File: skipole/ski/widgets/lists.py


"""Contains widgets for lists"""

from .. import tag
from . import Widget, ClosedWidget, FieldArg, FieldArgList, FieldArgTable, FieldArgDict, AnchorClickEventMixin


class UList1(Widget):
    """An unordered list widget."""

    # This class does not display any error messages
    display_errors = False

    arg_descriptions = {
                       'highlight_class':FieldArg("cssclass", ''),
                       'set_highlight':FieldArg("boolean", False, jsonset=True),
                       'contents': FieldArgList('text', jsonset=True),
                       'even_class':FieldArg("cssclass", ""),
                       'odd_class':FieldArg("cssclass", "")
                       }

    def __init__(self, name=None, brief='', **field_args):
        """
        highlight_class: CSS class will replace widget_class if set_highlight is True
        set_highlight: Set to True to force highlight_class to replace widget_class
                       Set to False to remove highlight_class and return to widget_class
        contents: List of text strings to be shown, gives the number of list items
        even_class: class of even li elements, if empty string, then no class will be applied
        odd_class: class of odd li elements, if empty string, then no class will be applied
        """
        Widget.__init__(self, name=name, brief=brief, **field_args)
        self.tag_name = "ul"


    def _build(self, page, ident_list, environ, call_data, lang):
        "Build the list"
        contents = self.wf.contents

        # any label:value added to self.jlabels will be set in a javascript fieldvalues attribute for the widget
        if self.wf.highlight_class:
            self.jlabels['highlight_class'] = self.wf.highlight_class
            if self.wf.set_highlight:
                self.attribs["class"] = self.wf.highlight_class

        # set even li class
        if self.wf.even_class:
            _even = self.wf.even_class
            self.jlabels['even_class'] = _even
        else:
            _even = ''
        # set odd li class
        if self.wf.odd_class:
            _odd = self.wf.odd_class
            self.jlabels['odd_class'] = _odd
        else:
            _odd = ''
        for index, item in enumerate(contents):
            if _even and (index % 2):
                self[index] = tag.Part(tag_name="li", attribs={"class":_even}, text=item)
            elif _odd and not (index % 2):
                self[index] = tag.Part(tag_name='li', attribs={"class":_odd}, text=item)
            else:
                self[index] = tag.Part(tag_name='li', text=item)


    @classmethod
    def description(cls):
        """Returns a text string to illustrate the widget"""
        return """<ul>  <!-- with widget id and class widget_class -->
  <li> <!-- with style from odd_class or even_class -->
      <!-- with text from each contents element -->
  </li>
  <!-- with appropriate number of list elements -->
</ul>
"""


class UList2(Widget):
    """An unordered list widget containing html"""

    # This class does not display any error messages
    display_errors = False

    arg_descriptions = {
                       'highlight_class':FieldArg("cssclass", ''),
                       'set_highlight':FieldArg("boolean", False, jsonset=True),
                       'set_html': FieldArgList('text', jsonset=True),
                       'even_class':FieldArg("cssclass", ""),
                       'odd_class':FieldArg("cssclass", "")
                       }

    def __init__(self, name=None, brief='', **field_args):
        """
        highlight_class: CSS class will replace widget_class if set_highlight is True
        set_highlight: Set to True to force highlight_class to replace widget_class
                       Set to False to remove highlight_class and return to widget_class
        set_html: List of html strings to be shown, html will be shown unescaped
        even_class: class of even li elements, if empty string, then no class will be applied
        odd_class: class of odd li elements, if empty string, then no class will be applied
        """
        Widget.__init__(self, name=name, brief=brief, **field_args)
        self.tag_name = "ul"

    def _build(self, page, ident_list, environ, call_data, lang):
        "Build the list"
        contents = self.wf.set_html

        # any label:value added to self.jlabels will be set in a javascript fieldvalues attribute for the widget

        if self.wf.highlight_class:
            self.jlabels['highlight_class'] = self.wf.highlight_class
            if self.wf.set_highlight:
                self.attribs["class"] = self.wf.highlight_class

        # set even li class
        if self.wf.even_class:
            _even = self.wf.even_class
            self.jlabels['even_class'] = _even
        else:
            _even = ''
        # set odd li class
        if self.wf.odd_class:
            _odd = self.wf.odd_class
            self.jlabels['odd_class'] = _odd
        else:
            _odd = ''

        for index, item in enumerate(contents):
            if _even and (index % 2):
                self[index] = tag.Part(tag_name="li", attribs={"class":_even}, text=item)
            elif _odd and not (index % 2):
                self[index] = tag.Part(tag_name='li', attribs={"class":_odd}, text=item)
            else:
                self[index] = tag.Part(tag_name='li', text=item)
            self[index].htmlescaped = False
            self[index].linebreaks=False


    @classmethod
    def description(cls):
        """Returns a text string to illustrate the widget"""
        return """<ul>  <!-- with widget id and class widget_class -->
  <li> <!-- with style from odd_class or even_class -->
      <!-- with html from each set_html element -->
  </li>
  <!-- with appropriate number of list elements -->
</ul>
"""


class TableList(AnchorClickEventMixin, Widget):
    """A table of a single text column, followed by Up, Down and Remove button link columns
       Each link having an individual get field set
       Typically used to display and order a Python list"""


    # This class does not display any error messages
    display_errors = False

    arg_descriptions = {
                        'highlight_class':FieldArg("cssclass", ''),
                        'set_highlight':FieldArg("boolean", False, jsonset=True),
                        'even_class':FieldArg("cssclass", ""),
                        'odd_class':FieldArg("cssclass", ""),
                        'maximize_text_col':FieldArg('boolean',True),
                        'hide':FieldArg("boolean", False, jsonset=True),
                        'button_class':FieldArg("cssclass", ""),
                        'remove_button_text':FieldArg('text', 'Remove'),
                        'up_link_ident':FieldArg("url", 'no_javascript'),
                        'up_json_ident':FieldArg("url", ''),
                        'down_link_ident':FieldArg("url", 'no_javascript'),
                        'down_json_ident':FieldArg("url", ''),
                        'remove_link_ident':FieldArg("url", 'no_javascript'),
                        'remove_json_ident':FieldArg("url", ''),
                        'contents':FieldArgTable(['text', 'text', 'text', 'text'], valdt=True, jsonset=True)
                        }

    def __init__(self, name=None, brief='', **field_args):
        """
        highlight_class: CSS class will replace widget_class if set_highlight is True
        set_highlight: Set to True to force highlight_class to replace widget_class
                       Set to False to remove highlight_class and return to widget_class
        even_class: class of even rows, if empty string, then no class will be applied
        odd_class: class of odd rows, if empty string, then no class will be applied
        maximize_text_col: If True the text column is made large, with the button columns reduced to the size of the buttons
        hide: if True, table will be hidden
        button_class: The CSS class to apply to the buttons
        remove_button_text: Text on the Remove button
        up_link_ident: For the up arrow button: the target page link ident, label or url, if javascript not available
        up_json_ident: For the up arrow button: the url, ident or label to link, expecting a json file to be returned
        down_link_ident: For the down arrow button: the target page link ident, label or url, if javascript not available
        down_json_ident: For the down arrow button: the url, ident or label to link, expecting a json file to be returned
        remove_link_ident: For the Remove button: the target page link ident, label or url, if javascript not available
        remove_json_ident: For the Remove button: the url, ident or label to link, expecting a json file to be returned
        contents: col 0 is the text string to place in the first column,
                  col 1 is the get field contents of the up button link
                  col 2 is the get field contents of the down button link
                  col 3 is the get field contents of the remove button link
                  This fieldname used as the widgfield
        """
        Widget.__init__(self, name=name, brief=brief, **field_args)
        self.tag_name = "table"


    def _build(self, page, ident_list, environ, call_data, lang):
        "Build the table"
        # Hides widget if no error and hide is True
        self.widget_hide(self.wf.hide)
        fieldtable = self.wf.contents

        # any label:value added to self.jlabels will be set in a javascript fieldvalues attribute for the widget

        if self.wf.highlight_class:
            self.jlabels['highlight_class'] = self.wf.highlight_class
            if self.wf.set_highlight:
                self.attribs["class"] = self.wf.highlight_class
        self.jlabels['button_class'] = self.wf.button_class
        self.jlabels['remove_button_text'] = self.wf.remove_button_text

        # set even row colour
        if self.wf.even_class:
            even = self.wf.even_class
        else:
            even = ''
        self.jlabels['even_class'] = even

        # set odd row colour
        if self.wf.odd_class:
            odd = self.wf.odd_class
        else:
            odd = ''
        self.jlabels['odd_class'] = odd

        if self.wf.up_json_ident:
            self.jlabels['up_json_url'] = self.get_url(self.wf.up_json_ident)
        if self.wf.down_json_ident:
            self.jlabels['down_json_url'] = self.get_url(self.wf.down_json_ident)
        if self.wf.remove_json_ident:
            self.jlabels['remove_json_url'] = self.get_url(self.wf.remove_json_ident)

        if self.wf.up_link_ident:
            up_url = self.get_url(self.wf.up_link_ident)
            self.jlabels['up_link_url'] = up_url
        else:
            up_url = ''

        if self.wf.down_link_ident:
            down_url = self.get_url(self.wf.down_link_ident)
            self.jlabels['down_link_url'] = down_url
        else:
            down_url = ''

        if self.wf.remove_link_ident:
            remove_url = self.get_url(self.wf.remove_link_ident)
            self.jlabels['remove_link_url'] = remove_url
        else:
            remove_url = ''

        if self.wf.maximize_text_col:
           self.jlabels['maximize_text_col'] = True

        if not fieldtable:
            return
        len_fieldtable = len(fieldtable)
        lastrow = len_fieldtable - 1

        # create rows
        for rownumber, row in enumerate(fieldtable):
            if even and (rownumber % 2) :
                self[rownumber] = tag.Part(tag_name="tr", attribs={"class":even})
            elif odd and not (rownumber % 2):
                self[rownumber] = tag.Part(tag_name='tr', attribs={"class":odd})
            else:
                self[rownumber] = tag.Part(tag_name='tr')

            # first column is text
            if self.wf.maximize_text_col:
                self[rownumber][0] = tag.Part(tag_name='td', text = row[0], attribs={"style":"width : 100%;"})
            else:
                self[rownumber][0] = tag.Part(tag_name='td', text = row[0])

            # Next column is up button link
            self[rownumber][1] = tag.Part(tag_name='td')
            if not rownumber:
                # rownumber 0 has no up button
                self[rownumber][1][0] = tag.HTMLSymbol("&nbsp;")
            else:
                if self.wf.button_class:
                    self[rownumber][1][0] = tag.Part(tag_name='a', attribs = {"role":"button", "class":self.wf.button_class})
                else:
                    self[rownumber][1][0] = tag.Part(tag_name='a', attribs = {"role":"button"})
                self[rownumber][1][0].htmlescaped=False
                if up_url:
                    self[rownumber][1][0][0] = tag.HTMLSymbol("&uarr;")
                    # create a url for the href
                    get_fields = {self.get_formname("contents"):row[1]}
                    self[rownumber][1][0].attribs["href"] = self.make_get_url(page, up_url, get_fields, True)
                else:
                    self[rownumber][1][0] = "Warning: broken link"

            # Next column is down button link
            self[rownumber][2] = tag.Part(tag_name='td')
            if rownumber == lastrow:
                # last row has no down button
                self[rownumber][2][0] = tag.HTMLSymbol("&nbsp;")
            else:
                if self.wf.button_class:
                    self[rownumber][2][0] = tag.Part(tag_name='a', attribs = {"role":"button", "class":self.wf.button_class})
                else:
                    self[rownumber][2][0] = tag.Part(tag_name='a', attribs = {"role":"button"})
                self[rownumber][2][0].htmlescaped=False
                if down_url:
                    self[rownumber][2][0][0] = tag.HTMLSymbol("&darr;")
                    # create a url for the href
                    get_fields = {self.get_formname("contents"):row[2]}
                    self[rownumber][2][0].attribs["href"] = self.make_get_url(page, down_url, get_fields, True)
                else:
                    self[rownumber][2][0] = "Warning: broken link"

            # Next column is Remove button link
            self[rownumber][3] = tag.Part(tag_name='td')
            if self.wf.button_class:
                self[rownumber][3][0] = tag.Part(tag_name='a', attribs = {"role":"button", "class":self.wf.button_class})
            else:
                self[rownumber][3][0] = tag.Part(tag_name='a', attribs = {"role":"button"})
            self[rownumber][3][0].htmlescaped=False
            if remove_url:
                if self.wf.remove_button_text:
                    self[rownumber][3][0][0] = self.wf.remove_button_text
                else:
                    self[rownumber][3][0][0] = "Remove"
                # create a url for the href
                get_fields = {self.get_formname("contents"):row[3]}
                self[rownumber][3][0].attribs["href"] = self.make_get_url(page, remove_url, get_fields, True)
            else:
                self[rownumber][3][0] = "Warning: broken link"



    @classmethod
    def description(cls):
        """Returns a text string to illustrate the widget"""
        return """
<table>  <!-- with widget id and class widget_class -->
                   <!-- and attribute style=display:none if hide is True -->
  <tr> <!-- with class from even or odd classes -->
    <td> <!-- text string --> </td>
    <td>
      <a role="button" href="#">
      <!-- With class set by button_class, and the href link will be the url of the up_link_ident with get field -->
        &uarr;
      </a>
    </td>
    <td>
      <a role="button" href="#">
      <!-- With class set by button_class, and the href link will be the url of the down_link_ident with get field -->
        &darr;
      </a>
    </td>
    <td>
      <a role="button" href="#">
      <!-- With class set by button_class, and the href link will be the url of the remove_link_ident with get field -->
        Remove
      </a>
    </td>
  </tr>
  <!-- rows repeated -->
</table>"""

File: skipole/skis/static/js/lists.js


SKIPOLE.lists = {};

SKIPOLE.lists.UList1 = function (widg_id, error_message, fieldmap) {
    SKIPOLE.BaseWidget.call(this, widg_id, error_message, fieldmap);
    this.display_errors = false;
    };
SKIPOLE.lists.UList1.prototype = Object.create(SKIPOLE.BaseWidget.prototype);
SKIPOLE.lists.UList1.prototype.constructor = SKIPOLE.lists.UList1;
SKIPOLE.lists.UList1.prototype.setvalues = function (fieldlist, result) {
    if (!this.widg_id) {
        return;
        }
    let the_widg = this.widg;
    let fieldvalues = this.fieldvalues;
    // contents
    let contents = this.fieldarg_in_result('contents', result, fieldlist);
    // the class of the elements
    let even_class = fieldvalues["even_class"];
    let odd_class = fieldvalues["odd_class"];

    if (contents) {
        // If a request to renew the list is recieved, this block
        // empties the existing list, and re-draws it
        let rows = contents.length;
        // empty the list
        the_widg.empty();
        // and now start filling it again
        let htmlcontent = "";
        for (row = 0; row < rows; row++) {
            // for each row
            if (even_class && (row % 2)) {
                htmlcontent += "<li class = \"" + even_class + "\">";
                }
            else if (odd_class && (!(row % 2))) {
                htmlcontent += "<li class = \"" + odd_class + "\">";
                }
            else {
                htmlcontent += "<li>";
                }
            htmlcontent += SKIPOLE.textbr(contents[row]);
            // close the cell
            htmlcontent += "</li>";
            }
        the_widg.html(htmlcontent);
        }
    };



SKIPOLE.lists.UList2 = function (widg_id, error_message, fieldmap) {
    SKIPOLE.BaseWidget.call(this, widg_id, error_message, fieldmap);
    this.display_errors = false;
    };
SKIPOLE.lists.UList2.prototype = Object.create(SKIPOLE.BaseWidget.prototype);
SKIPOLE.lists.UList2.prototype.constructor = SKIPOLE.lists.UList2;
SKIPOLE.lists.UList2.prototype.setvalues = function (fieldlist, result) {
    if (!this.widg_id) {
        return;
        }
    let the_widg = this.widg;
    let fieldvalues = this.fieldvalues;
    // contents
    let contents = this.fieldarg_in_result('set_html', result, fieldlist);
    // the class of the elements
    let even_class = fieldvalues["even_class"];
    let odd_class = fieldvalues["odd_class"];

    if (contents) {
        // If a request to renew the list is recieved, this block
        // empties the existing list, and re-draws it
        let rows = contents.length;
        // empty the list
        the_widg.empty();
        // and now start filling it again
        let htmlcontent = "";
        for (row = 0; row < rows; row++) {
            // for each row
            if (even_class && (row % 2)) {
                htmlcontent += "<li class = \"" + even_class + "\">";
                }
            else if (odd_class && (!(row % 2))) {
                htmlcontent += "<li class = \"" + odd_class + "\">";
                }
            else {
                htmlcontent += "<li>";
                }
            htmlcontent += contents[row];
            // close the cell
            htmlcontent += "</li>";
            }
        the_widg.html(htmlcontent);
        }
    };


SKIPOLE.lists.TableList = function (widg_id, error_message, fieldmap) {
    SKIPOLE.BaseWidget.call(this, widg_id, error_message, fieldmap);
    this.display_errors = false;
    };
SKIPOLE.lists.TableList.prototype = Object.create(SKIPOLE.BaseWidget.prototype);
SKIPOLE.lists.TableList.prototype.constructor = SKIPOLE.lists.TableList;
SKIPOLE.lists.TableList.prototype.setvalues = function (fieldlist, result) {
   if (!this.widg_id) {
        return;
        }
    let widg_id = this.widg_id
    let the_widg = this.widg;
    let fieldvalues = this.fieldvalues;
    // hide the widget
    let set_hide = this.fieldarg_in_result('hide', result, fieldlist);
    if (set_hide != undefined) {
        if (set_hide) {
            if (the_widg.is(":visible")) {
                the_widg.fadeOut('slow');
                }
            }
        else {
            if (!(the_widg.is(":visible"))) {
                the_widg.fadeIn('slow');
                 }
            }
        }

    // If a request to renew the table is recieved
    // empty the existing table, and re-draw it

    // The table contents
    let contents = this.fieldarg_in_result('contents', result, fieldlist);
    if (contents == undefined){
       return;
       }

    // the class of the button's if any
    let button_class = fieldvalues["button_class"];
    let remove_button_text = fieldvalues["remove_button_text"];
    if (!remove_button_text) {
          remove_button_text = "Remove"
          }

    // the class of the rows
    let even_class = fieldvalues["even_class"];
    let odd_class = fieldvalues["odd_class"];

    // button urls
    let up_link_url = fieldvalues["up_link_url"];
    let down_link_url = fieldvalues["down_link_url"];
    let remove_link_url = fieldvalues["remove_link_url"];

    let maximize_text_col = fieldvalues["maximize_text_col"];

    // empty the table
    the_widg.empty();

    // The table contents
    if (contents) {
        let rows = contents.length;
        // Start filling the empty table
        let htmlcontent = "";
        for (row = 0; row < rows; row++) {
            // for each row in the table
            // row class
            if (even_class && (row % 2)) {
                htmlcontent += "<tr class = \"" + even_class + "\">";
                }
            else if (odd_class && (!(row % 2))) {
                htmlcontent += "<tr class = \"" + odd_class + "\">";
                }
            else {
                htmlcontent += "<tr>";
                }

            // for each row, build the td's from content
      
            let tablerow = contents[row];
            // tablerow is [textitem, upget, downget, removeget]

            // first column is text
            if (maximize_text_col) {
                htmlcontent += "<td style=\"width : 100%;\">";
                }
            else {
                htmlcontent += "<td>";
                }
            htmlcontent += SKIPOLE.textbr(tablerow[0]) + "</td>";

            // 2nd column is uparrow
           if (!row) {
               // row zero has no up button
               htmlcontent += "<td>&nbsp;</td>";
               }
           else if (!up_link_url) {
               htmlcontent += "<td>Warning: broken link</td>";
               }
           else {
                // its a link, apply button class
                htmlcontent += "<td>"
                let url = up_link_url;
                if (button_class) {
                    htmlcontent +=  "<a role=\"button\" class=\"" + button_class + "\"";
                    }
                else {
                    htmlcontent +=  "<a role=\"button\"";
                    }
                // get url and create href attribute
                if (tablerow[1]) {
                    url += "?ident=" + SKIPOLE.identdata + "&" + this.formname("contents") + "=" + tablerow[1];
                    }
                else {
                    url += "?ident=" + SKIPOLE.identdata;
                    }
                htmlcontent +=  " href=\"" + url + "\">&uarr;</a></td>";
                }


            // 3rd column is downarrow
           if (row == rows-1) {
               // last row has no down button
               htmlcontent += "<td>&nbsp;</td>";
               }
           else if (!down_link_url) {
               htmlcontent += "<td>Warning: broken link</td>";
               }
           else {
                // its a link, apply button class
                htmlcontent += "<td>"
                let url = down_link_url;
                if (button_class) {
                    htmlcontent +=  "<a role=\"button\" class=\"" + button_class + "\"";
                    }
                else {
                    htmlcontent +=  "<a role=\"button\"";
                    }
                // get url and create href attribute
                if (tablerow[2]) {
                    url += "?ident=" + SKIPOLE.identdata + "&" + this.formname("contents") + "=" + tablerow[2];
                    }
                else {
                    url += "?ident=" + SKIPOLE.identdata;
                    }
                htmlcontent +=  " href=\"" + url + "\">&darr;</a></td>";
                }

            // 4th column is Remove
           if (!remove_link_url) {
               htmlcontent += "<td>Warning: broken link</td>";
               }
           else {
                // its a link, apply button class
                htmlcontent += "<td>"
                let url = remove_link_url;
                if (button_class) {
                    htmlcontent +=  "<a role=\"button\" class=\"" + button_class + "\"";
                    }
                else {
                    htmlcontent +=  "<a role=\"button\"";
                    }
                // get url and create href attribute
                if (tablerow[3]) {
                    url += "?ident=" + SKIPOLE.identdata + "&" + this.formname("contents") + "=" + tablerow[3];
                    }
                else {
                    url += "?ident=" + SKIPOLE.identdata;
                    }
                htmlcontent +=  " href=\"" + url + "\">" + remove_button_text + "</a></td>";
                }

            // close the row
            htmlcontent += "</tr>";
            }
        the_widg.html(htmlcontent);
        // as table was emptied, a new click event has to be applied to the buttons
        $("#" + widg_id + " a").click(function (e) {
              SKIPOLE.widgets[widg_id].eventfunc(e);
              });
        }

    };

SKIPOLE.lists.TableList.prototype.eventfunc = function (e) {
    SKIPOLE.skiprefresh = true;
    if (!this.widg_id) {
        return;
        }
    let fieldvalues = this.fieldvalues;
    let button = $(e.target);
    let href = button.attr('href');
    if (!href) {
        return;
        }

    let senddata = href.substring(href.indexOf('?')+1);

    /* col 0 is text, no link
       col 1 calls up_json_url,
       col 2 calls down_json_url,
       col 3 calls remove_json_url

       So need to find the column of the target to find which json url to call */

    let col = button.parent().index();
    let json_url = "";

    if (col == 1) {
                   json_url = fieldvalues["up_json_url"];
                  }
    else if (col == 2) {
                   json_url = fieldvalues["down_json_url"];
                  }
    else if (col == 3) {
                   json_url = fieldvalues["remove_json_url"];
                       }
    else {
          return;
         }

    if (!json_url) {
          return;
          }

    // so we have a json_url to call, and senddata has the query string

    e.preventDefault();
    // respond to json or html
    $.ajax({
          url: json_url,
          data: senddata
              })
          .done(function(result, textStatus, jqXHR) {
             if (jqXHR.responseJSON) {
                  // JSON response
                  SKIPOLE.setfields(result);
                  } else {
                      // html response
                      document.open();
                      document.write(result);
                      document.close();
                      }
              })
          .fail(function( jqXHR, textStatus, errorThrown ) {
                      if (jqXHR.status == 400 || jqXHR.status == 404 || jqXHR.status == 500)  {
                          document.open();
                          document.write(jqXHR.responseText);
                          document.close();
                          }
                      else {
                          SKIPOLE.json_failed( jqXHR, textStatus, errorThrown );
                          }
              });
    };




